test/pkcs12_format_test.c

```diff
@@ -0,0 +1,881 @@
+                    if (!TEST_true(check_asn1_string(av, attrs[k].value)))
+                        return 0;
+                }
+                break;
+            }
+        }
+    }
+    return 1;
+}
+
+int check_certbag(const PKCS12_SAFEBAG *bag, const char *cert_pem, PKCS12_ATTR *attrs, int num_attrs) {
+    X509 *x509 = NULL;
+    X509 *ref_x509 = NULL;
+
+    if (!check_attrs(PKCS12_SAFEBAG_get0_attrs(bag), attrs, num_attrs))
+        return 0;
+
+    if (!TEST_int_eq(PKCS12_SAFEBAG_get_nid(bag), NID_certBag))
+        return 0;
+
+    if (!TEST_int_eq(PKCS12_SAFEBAG_get_bag_nid(bag), NID_x509Certificate))
+        return 0;
+
+    x509 = PKCS12_SAFEBAG_get1_cert(bag);
+    if (!TEST_ptr(x509))
+        return 0;
+
+    ref_x509 = load_cert(cert_pem);
+    if (!TEST_false(X509_cmp(x509, ref_x509)))
+        return 0;
+
+    X509_free(x509);
+    X509_free(ref_x509);
+    return 1;
+}
+
+int check_keybag(const PKCS12_SAFEBAG *bag, const char *key_pem, PKCS12_ATTR *attrs, int num_attrs, PKCS12_ENC *enc) {
+    EVP_PKEY *pkey;
+    EVP_PKEY *ref_pkey;
+    PKCS8_PRIV_KEY_INFO *p8;
+    const PKCS8_PRIV_KEY_INFO *p8c;
+
+    if (!check_attrs(PKCS12_SAFEBAG_get0_attrs(bag), attrs, num_attrs))
+        return 0;
+
+    switch (PKCS12_SAFEBAG_get_nid(bag)) {
+    case NID_keyBag:
+        p8c = PKCS12_SAFEBAG_get0_p8inf(bag);
+        if (!TEST_ptr(pkey = EVP_PKCS82PKEY(p8c)))
+            return 0;
+        /* TODO: handle key attributes */
+        //print_attribs(out, PKCS8_pkey_get0_attrs(p8c), "Key Attributes");
+        break;
+
+    case NID_pkcs8ShroudedKeyBag:
+        if (!TEST_ptr(p8 = PKCS12_decrypt_skey(bag, enc->pass, strlen(enc->pass))))
+            return 0;
+        if (!TEST_ptr(pkey = EVP_PKCS82PKEY(p8))) {
+            PKCS8_PRIV_KEY_INFO_free(p8);
+            return 0;
+        }
+        /* TODO: handle key attributes */
+        //print_attribs(out, PKCS8_pkey_get0_attrs(p8), "Key Attributes");
+        PKCS8_PRIV_KEY_INFO_free(p8);
+        break;
+
+    default:
+        return 0;
+    }
+
+    /* PKEY compare returns 1 for match */
+    ref_pkey = load_pkey(key_pem);
+    if (!TEST_true(EVP_PKEY_cmp(pkey, ref_pkey)))
+        return 0;
+
+    EVP_PKEY_free(pkey);
+    EVP_PKEY_free(ref_pkey);
+    return 1;
+}
+
+int check_secretbag(const PKCS12_SAFEBAG *bag, int secret_nid, const char *secret, PKCS12_ATTR *attrs, int num_attrs) {
+    const ASN1_TYPE *at = NULL;
+
+    if (!check_attrs(PKCS12_SAFEBAG_get0_attrs(bag), attrs, num_attrs))
+        return 0;
+
+    if (!TEST_int_eq(PKCS12_SAFEBAG_get_nid(bag), NID_secretBag))
+        return 0;
+
+    if (!TEST_int_eq(PKCS12_SAFEBAG_get_bag_nid(bag), secret_nid))
+        return 0;
+
+    if (!TEST_true(check_asn1_string(PKCS12_SAFEBAG_get0_bag_obj(bag), secret)))
+        return 0;
+
+    return 1;
+}
+
+/* --------------------------------------------------------------------------
+ * PKCS12 builder macros
+ */
+
+#define start_pkcs12(fname)  \
```

Comment:
I suspect an opening `{` is better here to avoid declarations mixed with code.

Reply:
The BIO declaration here is outside the block so that later after the end_pkcs12() the p12bio can be read by the start_check_pkcs12() macro.

Reply:
What about starting a block before the declaration:
```
#define start_pkcs12(fname)  \
    { \
        BIO *p12bio = NULL; \
```